umh-core/pkg/fsm/nmap/reconcile.go

```diff
@@ -0,0 +1,444 @@
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentNmapMonitor, n.baseFSMInstance.GetID()+".reconcileTransitionToStopped", time.Since(start))
+	}()
+
+	// If not yet stopped and not already stopping, initiate stop.
+	if currentState != OperationalStateStopped && currentState != OperationalStateStopping {
+		if err := n.monitoringStop(ctx); err != nil {
+			return err, true
+		}
+		return n.baseFSMInstance.SendEvent(ctx, EventStop), true
+	}
+
+	// If currently stopping, verify that it is now fully stopped.
+	if currentState == OperationalStateStopping {
+		if n.IsStopped() {
+			return n.baseFSMInstance.SendEvent(ctx, EventStopDone), true
+		}
+		return nil, false
+	}
+
+	return nil, false
+}
+
+// reconcileStartingState handles the various starting phase states when transitioning to Active.
+func (n *NmapInstance) reconcileStartingState(ctx context.Context, filesystemService filesystem.Service, currentState string, currentTime time.Time) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentBenthosInstance, n.baseFSMInstance.GetID()+".reconcileStartingState", time.Since(start))
+	}()
+
+	switch currentState {
+	case OperationalStateStarting:
+		return n.baseFSMInstance.SendEvent(ctx, EventStartDone), true
+	default:
+		return fmt.Errorf("invalid starting state: %s", currentState), false
+	}
+}
+
+// reconcileRunningState handles the various running states when transitioning to Active.
+func (n *NmapInstance) reconcileRunningState(ctx context.Context, filesystemService filesystem.Service, currentState string, currentTime time.Time) (err error, reconciled bool) {
+	start := time.Now()
+	defer func() {
+		metrics.ObserveReconcileTime(metrics.ComponentBenthosInstance, n.baseFSMInstance.GetID()+".reconcileRunningState", time.Since(start))
+	}()
+
+	switch currentState {
+	case OperationalStateOpen:
+		// If we're in open, we need to check whether it is degraded
+		if !n.isNmapHealthy() {
+			return n.baseFSMInstance.SendEvent(ctx, EventNmapExecutionFailed), true
+		}
+
+		changedPortState, event := n.checkPortState(currentState)
+		if changedPortState {
+			return n.baseFSMInstance.SendEvent(ctx, event), true
+		}
+
+		return nil, false
+	case OperationalStateFiltered:
+		// If we're in filtered, we need to check whether it is degraded
+		if !n.isNmapHealthy() {
+			return n.baseFSMInstance.SendEvent(ctx, EventNmapExecutionFailed), true
+		}
+
+		changedPortState, event := n.checkPortState(currentState)
+		if changedPortState {
+			return n.baseFSMInstance.SendEvent(ctx, event), true
+		}
+
+		return nil, false
+	case OperationalStateClosed:
+		// If we're in closed, we need to check whether it is degraded
+		if !n.isNmapHealthy() {
+			return n.baseFSMInstance.SendEvent(ctx, EventNmapExecutionFailed), true
+		}
+
+		changedPortState, event := n.checkPortState(currentState)
+		if changedPortState {
+			return n.baseFSMInstance.SendEvent(ctx, event), true
```

Comment:
maybe as an idea: set reconciled to false if it switches between up/closed/filtered as there is not much happening here in the first place.

background: reconciled tells the control loop to stop executing at this point and wait for the next tick. it makes sense when we do "large actions" such as the creation of a new service etc. to give the system enough time to react, and it also makes sense for the starting, stopping, stopped, etc. phases for nmap, but maybe for up/filtered/closed it could make sense to keep reconciling (otherwise a single connection that went bad will now cause the entire loop to be skipped). could also do this for degraded state as well.

if we do it, we should definitely document that.